datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -1277,7 +1404,84 @@ pub(crate) fn from_substrait_literal(lit: &Literal) -> Result<ScalarValue> {
                 s,
             )
         }
-        Some(LiteralType::Null(ntype)) => from_substrait_null(ntype)?,
+        Some(LiteralType::List(l)) => {
+            let element_dt: Option<&DataType> = match dt {
+                Some(DataType::List(l)) => Ok(Some(l.data_type())),
```

Comment:
I am wondering if `dt` is redundant here and can be removed. We still need to check whether it is a `LargeList`, just as we check `type_variation_reference` below.

Is it possible to only extract `element_dt` from `elements[0]`?  I guess that `elements` are not empty because substrait has a special type called EmptyList.
```rust
 pub struct List {
            /// A homogeneously-typed list of one or more expressions that form the
            /// list entries. To specify an empty list, use Literal.empty_list
            /// (otherwise type information would be missing).
            #[prost(message, repeated, tag="1")]
            pub values: ::prost::alloc::vec::Vec<super::super::Expression>,
        }
```

Reply:
Yeah, we already do that (look at the first element) if `dt` is not given. I think the reason I included `dt` was because Substrait types don't include the field names for structs, so not using `dt` would mean using something like `names` and `name_idx` instead (like in from_substrait_type_with_names).

Reply:
Thank you for your explanation @Blizzara . My concern is that using two types for List simultaneously is a bit verbose unless it is necessary.

I tested the following two cases: the first one panicked, and the second one also failed.
```rust

#[tokio::test]
async fn roundtrip_list_iteral() -> Result<()> {
    roundtrip("SELECT [] from data").await
}

#[tokio::test]
async fn roundtrip_struct_iteral() -> Result<()> {
    roundtrip("select struct(1 as name0, 3.14 as name1, 'e', true as name3) from data")
        .await
}
``` 
```sh
---- cases::roundtrip_logical_plan::roundtrip_list_iteral stdout ----
thread 'cases::roundtrip_logical_plan::roundtrip_list_iteral' panicked at consumer.rs:1425:41:
index out of bounds: the len is 0 but the index is 0
```
I think lists and structs are more complex than the virtual table, so it's better for them to have separate PRs.

Reply:
This was because I wasn't handling empty lists properly (producer created Lists rather than EmptyList) - that was fixed in the Lists PR. Thanks for catching!